codalab/lib/cli_util.py

```diff
@@ -71,11 +72,19 @@ def parse_key_target(spec):
         - <key>: (<key> if present,
                     empty string if ':' in spec but no <key>,
                     None otherwise)
-        - <value> (where value is everyhing after a <key>: (or everything if no key specified)
+        - <value> (where value is everything after a <key>: (or everything if no key specified)
     """
-
     match = re.match(TARGET_KEY_REGEX, spec)
-    return match.groups()
+    key, value = match.groups()
+    # This check covers three usage errors:
+    # 1. both key and value are empty, e.g. ":"
+    # 2. key is not empty, value is empty, e.g. "<key>:"
+    if key is not None and value == '':
```

Comment:
Isn't the right logic simply just that `value == ''` regardless of what `key` is?

Reply:
Yes, I originally thought the same, but once I saw these [lines](https://github.com/codalab/codalab-worksheets/blob/master/codalab/lib/bundle_cli.py#L1451-L1453), I was thinking maybe I should add a `None` value check. It seems that's the special case it looks for. But let me know if I misunderstood anything.

Reply:
How are those lines related to this check?  The current code would accept `key = None and value == ''`, which presumably can't happen anyway, but `value == ''` is simpler.  I'd just make sure we have tests for all the different cases.

Reply:
From those [lines](https://github.com/codalab/codalab-worksheets/blob/master/codalab/lib/bundle_cli.py#L1451-L1453), it looks like it accepts `None` value and will convert it to an empty string when `None` value passed in. If this is something won't happen, maybe I should delete the check in bundle_cli.py to make things more clear.

Reply:
It's worth following how `targets` gets used downstream...it would be nice not to have a `''` key in addition to `None`, but we shouldn't break anything.  I think there are three (valid cases):
1) `key:value`
2) `:value`
3) `value` (specific to make bundles)
We need to distinguish between 2) and 3).

Reply:
I am really not sure if we can get rid of `None` key and `''` key check here. The following three functions call `parse_key_target`:
1. in [resolve_key_targets](https://github.com/codalab/codalab-worksheets/blob/master/codalab/lib/bundle_cli.py#L571):https://github.com/codalab/codalab-worksheets/blob/fa36a533c25d9a24dea51bc92775f580c1a81155/codalab/lib/bundle_cli.py#L579-L584):
2. in [do_make_command](https://github.com/codalab/codalab-worksheets/blob/master/codalab/lib/bundle_cli.py#L1449) : https://github.com/codalab/codalab-worksheets/blob/fa36a533c25d9a24dea51bc92775f580c1a81155/codalab/lib/bundle_cli.py#L1453
It checks for the existence of `None` key for each target returned from item 1 (essentially, there will be at most 1 target having `None` key based on logic in item 1 ^).
3. in [desugar_command](https://github.com/codalab/codalab-worksheets/blob/fa36a533c25d9a24dea51bc92775f580c1a81155/codalab/lib/cli_util.py#L119-L125): https://github.com/codalab/codalab-worksheets/blob/fa36a533c25d9a24dea51bc92775f580c1a81155/codalab/lib/cli_util.py#L119-L125

It looks like each of them is parsing `''` key or `None` key to fulfilling their own purpose. If I misunderstand our scope here, please do let me know.

Reply:
Yeah, I agree we can't merge them.  It seems like `key == ''` corresponds to 2) and `key is None` coresponds to 3).  But the original change of `value == ''` seems safe.

Reply:
I see. I have updated the condition check.